bumble/profiles/aics.py

```diff
@@ -0,0 +1,482 @@
+# Constants
+# -----------------------------------------------------------------------------
+CHANGE_COUNTER_MAX_VALUE = 0xFF
+GAIN_SETTINGS_MIN_VALUE = 0
+GAIN_SETTINGS_MAX_VALUE = 255
+
+
+class ErrorCode(enum.IntEnum):
+    '''
+    Cf. 1.6 Application error codes
+    '''
+
+    INVALID_CHANGE_COUNTER = 0x80
+    OPCODE_NOT_SUPPORTED = 0x81
+    MUTE_DISABLED = 0x82
+    VALUE_OUT_OF_RANGE = 0x83
+    GAIN_MODE_CHANGE_NOT_ALLOWED = 0x84
+
+
+class Mute(enum.IntEnum):
+    '''
+    Cf. 2.2.1.2 Mute Field
+    '''
+
+    NOT_MUTED = 0x00
+    MUTED = 0x01
+    DISABLED = 0x02
+
+
+class GainMode(enum.IntEnum):
+    '''
+    Cf. 2.2.1.3 Gain Mode
+    '''
+
+    MANUAL_ONLY = 0x00
+    AUTOMATIC_ONLY = 0x01
+    MANUAL = 0x02
+    AUTOMATIC = 0x03
+
+
+class AudioInputStatus(enum.IntEnum):
+    '''
+    Cf. 3.4 Audio Input Status
+    '''
+
+    INATIVE = 0x00
+    ACTIVE = 0x01
+
+
+class AudioInputControlPointOpCode(enum.IntEnum):
+    '''
+    Cf. 3.5.1 Audio Input Control Point procedure requirements
+    '''
+
+    SET_GAIN_SETTING = 0x00
+    UNMUTE = 0x02
+    MUTE = 0x03
+    SET_MANUAL_GAIN_MODE = 0x04
+    SET_AUTOMATIC_GAIN_MODE = 0x05
+
+
+# -----------------------------------------------------------------------------
+class AudioInputState(Characteristic):
+    '''
+    Cf. 2.2.1 Audio Input State
+    '''
+
+    def __init__(
+        self,
+        uuid: Union[str, bytes, UUID],
+        properties: Characteristic.Properties,
+        permissions: Union[str, Attribute.Permissions],
+        gain_settings: int,
+        gain_settings_unit: int,
+        mute: Mute,
+        gain_mode: GainMode,
+        descriptors: Sequence[Descriptor] = (),
+    ):
+        self.gain_settings = gain_settings
+        self.gain_settings_unit = gain_settings_unit
+        self.mute: Mute = mute
+        self.gain_mode: GainMode = gain_mode
+        self.change_counter = 0
+
+        value = CharacteristicValue(read=self._on_read)
+
+        super().__init__(uuid, properties, permissions, value, descriptors)
+
+    def __bytes__(self) -> bytes:
+        return bytes(
+            [self.gain_settings, self.mute, self.gain_mode, self.change_counter]
+        )
+
+    def update_gain_settings_unit(self, gain_settings_unit: int) -> None:
+        self.gain_settings_unit = gain_settings_unit
+
+    def increment_gain_settings(self) -> None:
+        self.gain_settings += self.gain_settings_unit
+        self.increment_change_counter()
+
+    def decrement_gain_settings(self) -> None:
+        self.gain_settings -= self.gain_settings_unit
+        self.increment_change_counter()
+
+    def increment_change_counter(self):
+        self.change_counter += 0x01
+        if self.change_counter > CHANGE_COUNTER_MAX_VALUE:
+            self.change_counter = 0x00
+
+    async def notify_subscribers(self, connection: Connection) -> None:
+        await connection.device.notify_subscribers(
+            attribute=self.value, value=self.__bytes__()
```

Comment:
The usage of `__bytes__` is that when `bytes(obj)` is invoked, Python will try to find its `__bytes__` implementation and pass to it, and directly use of `__bytes__ ` is not very common in comparison to a `to_bytes()` stub. But I am not sure whether there is any guideline about the usage of `__bytes__`

Reply:
I wanted to use `__bytes__` directly to avoid intermediate calls. But apparently calling `bytes()` or `to_bytes()` is more appropriate, I'll go for this then 👍